NixOS/auto-upgrade: offer the possibility to define a reboot window during which the system may be automatically rebooted

r-vdp opened the pull request:
###### Motivation for this change
Some systems should not be rebooted at just any time. If the upgrade process takes too long, for instance because of a slow internet connection, or if the upgrade service is ran during production hours, we want to allow to define a window outside of which a reboot will not be performed.
The system will then reboot on the next run of the upgrade service which finishes inside the reboot window.

E.g. we can run the update service twice per week, once during the night and once during the day, but reboots are only allowed during the night. By doing so, a system that is usually shut down during the night will still receive updates and systems that are turned on 24/7 can be rebooted outside of production hours.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

R-VdP commented:
@danbst, could you review this?

R-VdP commented:
@danbst, thanks! Anything else needs to happen to have this merged?

danbst commented:
@R-VdP sorry, I wait for more :eyes: . But I don't quite know whom to ping for another review.
I'll redirect this to 20.03 release managers @NixOS/nixos-release-managers

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-are-your-goals-for-20-03/4773/14

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/92

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-20-03-feature-freeze/5655/29

rnhmjoj commented:
I haven't fully tried it yet but it looks all right.
I wrote a NixOS test for this but unfortunately it doesn't fully work: since the VM has no internet access Nix can't rebuild the system. I hoped it could pick derivations from the host nix store but it doesn't really work. If somebody wants to try fixing it it's here: https://github.com/rnhmjoj/nixpkgs/commit/01d12847c26904fed58629574f486cf0fe5b2044

infinisil reviewed (changes requested):
This doesn't show correct behavior when the time span goes through midnight. E.g. with `lower = "23:00"`, `upper = "02:00"` and let's say it runs on Tuesday 01:00, then the condition makes it only reboot if "Tuesday 01:00 is before Tuesday 02:00 and after Tuesday 23:00", which won't be the case.

r-vdp committed:
NixOS/auto-upgrade: offer the possibility to define a reboot window during which the system may be automatically rebooted

Some systems should not be rebooted at just any time. If the upgrade process takes too long, for instance because of a
slow internet connection, or if the upgrade service is ran during production hours, we want to allow to define a window
outside of which a reboot will not be performed.
The system will then reboot on the next run of the upgrade service which finishes inside the reboot window.

E.g. we can run the update service twice per week, once during the night and once during the day, but reboots are only
allowed during the night. By doing so, a system that is usually shut down during the night will still receive updates
and systems that are turned on 24/7 can be rebooted outside of production hours.

Co-authored-by: Silvan Mosberger <github@infinisil.com>

R-VdP commented:
Something went horribly wrong when pushing new changes, I guess I did something wrong while trying to rebase the commits. I did a force push to clear up the mess.

@Infinisil, I made a change to address your issue, we now only look at the hour independently of the day. Do you agree that this fixes the issue that you raised?

worldofpeace commented:
@R-VdP This happens to be a well known bug in GitHub for nixpkgs, we're pretty used to the "ping issue". Don't sweat it :smile:

Mic92 commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -91,24 +112,36 @@ let cfg = config.system.autoUpgrade; in
+          ${optionalString (cfg.rebootWindow != null) ''
+            elif [[ "''${current_time}" < "${cfg.rebootWindow.lower}" ]] || \
+                 [[ "''${current_time}" > "${cfg.rebootWindow.upper}" ]]; then
+              echo "Outside of configured reboot window, skipping."
```
Might be an overkill but a transient timer could just schedule the reboot into the next reboot window: https://wiki.archlinux.org/index.php/Systemd/Timers#Transient_.timer_units

Mic92 replied:
Or just `shutdown -r "$cfg.rebootWindow.lower"`.

R-VdP replied:
I guess we'd have to introduce a new option for that though. And maybe we'd need to make the days explicit too then, my servers only have a maintenance window  once per week, for instance, not every day...
We might be able to make the whole thing more flexible by integrating with systemd somehow instead of comparing strings ourselves, but I'd have to think a bit about how that could be done.

R-VdP replied:
Also, it kind of gets pushed into the next maintenance window already now, given that the service should run again then and still detect the changes between the current system and the new boot default.

Mic92 replied:
Ok. Sounds good enough to me than!

Atemu commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -91,24 +112,36 @@ let cfg = config.system.autoUpgrade; in
+        nixos-rebuild = "${config.system.build.nixos-rebuild}/bin/nixos-rebuild";
+        date = "/run/current-system/sw/bin/date";
+        readlink = "/run/current-system/sw/bin/readlink";
+        shutdown = "/run/current-system/sw/bin/shutdown";
```
I think it'd be better to make these paths relative to `pkgs.coreutils` and `pkgs.systemd`, not the current system.

R-VdP replied:
@Atemu: done.

JohnAZoidberg commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -91,24 +112,36 @@ let cfg = config.system.autoUpgrade; in
-          nixos-rebuild = "${config.system.build.nixos-rebuild}/bin/nixos-rebuild";
-        in
+        nixos-rebuild = "${config.system.build.nixos-rebuild}/bin/nixos-rebuild";
+        date     = "${pkgs.coreutils}/bin/date";
```
Shouldn't `date`, `readlink` and `shutdown` already be available, since `coreutils` is in the unit's `path`?

R-VdP replied:
I guess. I figured it wouldn't hurt to make the paths explicit, but I'm fine removing this if that's they way things are usually done.

R-VdP commented:
> 
> 
> This doesn't show correct behavior when the time span goes through midnight. E.g. with `lower = "23:00"`, `upper = "02:00"` and let's say it runs on Tuesday 01:00, then the condition makes it only reboot if "Tuesday 01:00 is before Tuesday 02:00 and after Tuesday 23:00", which won't be the case.

@Infinisil: Could you re-review and approve if I sufficiently addressed your comments? Thanks!

Infinisil commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -91,24 +112,36 @@ let cfg = config.system.autoUpgrade; in
-        '';
+          ${optionalString (cfg.rebootWindow != null) ''
+            elif [[ "''${current_time}" < "${cfg.rebootWindow.lower}" ]] || \
+                 [[ "''${current_time}" > "${cfg.rebootWindow.upper}" ]]; then
```
Still wrong. Example is `lower = 12:00`, `upper = 6:00`, `current = 18:00` -> This is within the window, but the code doesn't do it then.

After a bit of thinking, I'd say the best way to correctly implement this is to: Determine if `lower > upper`, and if so, flip them around. Then check whether `lower < cur < upper`, and toggle this result if `lower > upper` from before. If the result is true, do the reboot, otherwise don't.

Infinisil commented:
Taking another look at this, unless I'm missing something, I think there's a big flaw with this: It's not useful, because the upgrade time is already fixed by the `system.autoUpgrade.dates` options. Assuming `system.autoUpgrade.allowReboot` is enabled:
- If `system.autoUpgrade.dates` is within `system.autoUpgrade.rebootWindow`, then it always reboots at 4:40 if there was a kernel/initrd update.
- Otherwise it never does, even if there are kernel/initrd updates, because it always happens at the same time.

As such, the `rebootWindow` option doesn't really do anything. Yes there's `system.autoUpgrade.randomizedDelaySec`, but that doesn't make it much better.

R-VdP commented:
> Taking another look at this, unless I'm missing something, I think there's a big flaw with this: It's not useful, because the upgrade time is already fixed by the `system.autoUpgrade.dates` options. Assuming `system.autoUpgrade.allowReboot` is enabled:
> 
>     * If `system.autoUpgrade.dates` is within `system.autoUpgrade.rebootWindow`, then it always reboots at 4:40 if there was a kernel/initrd update.
> 
>     * Otherwise it never does, even if there are kernel/initrd updates, because it always happens at the same time.
> 
> 
> As such, the `rebootWindow` option doesn't really do anything. Yes there's `system.autoUpgrade.randomizedDelaySec`, but that doesn't make it much better.

@Infinisil my use case is that we have many servers that are in very remote locations on often slow uplinks (sometimes satellite). As such, when an upgrade requires downloading a lot of new store paths, it may not finish before working hours start again, and so I want to be able to make sure that under no circumstances our servers reboot during working hours.

We also have servers that are turned off outside of working hours, due to power restrictions. For those, we run the upgrade service during working hours when the servers are turned on, but I don't want them to reboot at that time, they will activate the new generation on the next morning when they are started again. 

Does that make sense?

Infinisil commented:
I see, that does make more sense. That does make me think of an alternative approach that could be a bit more deterministic:
- Have a timer which `nixos-rebuild build`'s the system every e.g. hour, and have it detect when there's a change (by comparing `/nix/store` hashes)
- Depending on whether an activation and a reboot is needed:
  - If neither is needed, don't do anything
  - If only an activation is needed, and we're within `system.autoUpgrade.activationWindow` (a new option), do an activation, otherwise don't do anything
  - If only a reboot is needed, and we're within `system.autoUpgrade.rebootWindow`, do a reboot, otherwise don't do anything
  - If both an activation and a reboot is needed:
    - If we're not within `activationWindow`, don't do anything (we can't reboot either, since we need to have the current system be activated first)
    - Otherwise, do an activation
      - If we're also within `rebootWindow`, do a reboot after activation

This would mean that you can control when both activation and reboots can happen, and if the script makes sure to gcroot the system derivation when it first builds it (even if it can't activate/reboot yet), you can be sure that it will do the activation/reboot the next time it can, since everything is built already. How does that sound?

In any case, I don't think this idea conflicts with this PR and can be implemented later.

Infinisil commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -96,6 +96,30 @@ in {
 
+      rebootWindow = mkOption {
+        description = ''
+          Define a lower and upper time value (in HH:MM format) which
+          constitute a time window during which reboots are allowed.
```
```suggestion
          Define a lower and upper time value (in HH:MM format) which
          constitute a time window during which reboots are allowed after an upgrade.
          This option only has an effect when `allowReboots` is enabled.
          The default value of `null` means that reboots are allowed all the time.
```

Infinisil replied:
Also, the `allowReboots` option should have a reference to the `rebootWindow` option like

> See `rebootWindow` for configuring the times at which a reboot is allowed.

R-VdP commented:
> I see, that does make more sense. That does make me think of an alternative approach that could be a bit more deterministic:
> 
>     * Have a timer which `nixos-rebuild build`'s the system every e.g. hour, and have it detect when there's a change (by comparing `/nix/store` hashes)
> 
>     * Depending on whether an activation and a reboot is needed:
>       
>       * If neither is needed, don't do anything
>       * If only an activation is needed, and we're within `system.autoUpgrade.activationWindow` (a new option), do an activation, otherwise don't do anything
>       * If only a reboot is needed, and we're within `system.autoUpgrade.rebootWindow`, do a reboot, otherwise don't do anything
>       * If both an activation and a reboot is needed:
>         
>         * If we're not within `activationWindow`, don't do anything (we can't reboot either, since we need to have the current system be activated first)
>         * Otherwise, do an activation
>           
>           * If we're also within `rebootWindow`, do a reboot after activation
> 
> 
> This would mean that you can control when both activation and reboots can happen, and if the script makes sure to gcroot the system derivation when it first builds it (even if it can't activate/reboot yet), you can be sure that it will do the activation/reboot the next time it can, since everything is built already. How does that sound?
> 
> In any case, I don't think this idea conflicts with this PR and can be implemented later.

That's an interesting suggestion indeed!

There is one issue that I see with this approach though. On our servers we make the difference between a deployment of a new config from our git repository (eg addition of a new user or an updated ssh key, updated definition of a service, etc) which never requires a reboot, and an actual upgrade. Only during an upgrade do we we also update the nix channel and so do we potentially need to reboot after.

Changes to our git repo are under certain conditions deployed automatically to our servers. However, if we upgrade every hour and then wait until we're in the reboot window to reboot, then the nix channel can already have been updated when a config change is being deployed and so at that point, if a reboot was required, we cannot easily activate this new config without also switching the kernel, restarting services, etc.

I also don't think that there's an easy way to roll back the nix channel to the point from which the current generation was derived. 

All of this would not necessarily be an issue, but I noticed in the past that activating a config with an updated kernel without rebooting can cause issues where we can for instance no longer exec into docker containers, which is quite annoying for our support staff. Besides the fact that restarting services during production hours can also cause issues. 

So, at least for our use case, it's important that we only do an upgrade when we'll actually be able to reboot the server.

When the upgrade service exceeds the reboot window, we obviously also end up in the described situation, but this is assumed to be quite exceptional since such long running updates are not common, usually only on a new release.

Infinisil commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -80,6 +80,7 @@ in {
           Reboot the system into the new generation instead of a switch
           if the new generation uses a different kernel, kernel modules
           or initrd than the booted system.
+          See `rebootWindow` for configuring the times at which a reboot is allowed.
```
NixOS options currently still need to use DocBook unfortunately :sweat_smile: 

```suggestion
          See <option>rebootWindow</option> for configuring the times at which a reboot is allowed.
```

Infinisil commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -96,6 +97,32 @@ in {
+        description = ''
+          Define a lower and upper time value (in HH:MM format) which
+          constitute a time window during which reboots are allowed after an upgrade.
+          This option only has an effect when `allowReboot` is enabled.
+          The default value of `null` means that reboots are allowed at any time.
```
```suggestion
          This option only has an effect when <option>allowReboot</option> is enabled.
          The default value of <literal>null</literal> means that reboots are allowed at any time.
```

infinisil reviewed (commented):
Sorry haven't had a lot of time recently, but I should have some now. This is the only nit that I think should still be addressed, after that I'll merge it :)

R-VdP commented:
@Infinisil great, thanks!

I incorporated the proposed changes, squashed everything together, and rebased the PR on top of master.

infinisil reviewed (approved):
Looking good, let's merge this finally!
